traversal/scanner/GraphIterator.java

```diff
@@ -176,82 +214,29 @@ private boolean computeAnswer() {
 
     private void traverse(ProcedureVertex<?, ?> procedureVertex) {
         VertexTraverser vertexTraverser = vertexTraversers.get(procedureVertex);
-        boolean verified = false;
-        Set<ProcedureVertex<?, ?>> verifyFailureCauses = new HashSet<>();
-        while (!verified && vertexTraverser.hasNextVertex()) {
-            vertexTraverser.takeNextVertex();
-            if (vertexTraverser.verifyLoops()
-                    && verifyTraversedScopes(procedureVertex, verifyFailureCauses)
-                    && verifyEdgeLookahead(procedureVertex, verifyFailureCauses)) {
-                verified = true;
-            } else {
-                vertexTraverser.clearCurrentVertex();
-                removeFromSuccessors(procedureVertex);
-            }
-        }
-        toRevisit.addAll(verifyFailureCauses);
-        if (verified) procedureVertex.outs().forEach(edge -> toTraverse.add(edge.to()));
-        else failed(procedureVertex);
+        if (vertexTraverser.findNextVertex()) success(vertexTraverser);
+        else failed(vertexTraverser);
     }
 
-    private void removeFromSuccessors(ProcedureVertex<?, ?> procedureVertex) {
-        procedureVertex.outs().forEach(edge -> {
-            vertexTraversers.get(edge.to()).removeEdgeInput(edge);
-            removeFromSuccessors(edge.to());
-        });
+    private void success(VertexTraverser vertexTraverser) {
+        iterate(vertexTraverser.procedureVertex.dependents()).link(iterate(vertexTraverser.implicitDependents))
+                .forEachRemaining(v -> {
+                    toTraverse.add(v);
+                    vertexTraversers.get(v).clear();
+                });
     }
 
-    private void failed(ProcedureVertex<?, ?> procedureVertex) {
-        procedureVertex.ins().forEach(edge -> toRevisit.add(edge.from()));
-        vertexTraversers.get(procedureVertex).reset();
-        toTraverse.add(procedureVertex);
+    private void failed(VertexTraverser vertexTraverser) {
```

Comment:
on failure to find yet another vertex, we note all the predecessors that could have caused the failure (eg. dependees and implicit dependees) in the `toRevisit` set and backtrack